### DISCUSSION TITLE
re-install of brew doesn't recognise previously installed packages?

### DISCUSSION THREAD

**Discussion by aviexk** [Homebrew/discussions#2580] [Everyday usage]
1. install brew 
2. `brew install somePackage`
3. `brew list` returns `somePackage`
3. uninstall brew (for whatever reason) 
4. reinstall brew (for whatever reason) 
5. `brew list` returns nothing 

is there anyway to reindex the reinstalled brew so that it can find the existing/previously installed packages on the machine? or is `brew reinstall <each package>` one by one the only way?

**Comment by gromgit** [comment]
If you were using the [official uninstall script](https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh), it would've warned you right at the start:
```
Warning: This script will remove:
/Users/aho/Library/Caches/Homebrew/
/Users/aho/Library/Logs/Homebrew/
/usr/local/Caskroom/
/usr/local/Cellar/
/usr/local/bin/brew -> /usr/local/bin/brew
Are you sure you want to uninstall Homebrew? This will remove your installed packages! [y/N]
```
So no, all the files you previously installed are actually _deleted_, not hidden somewhere to be "reindexed" later.

Fortunately, `brew bundle dump` actually dumps a list of installed Homebrew packages in a `Brewfile` in the current directory, while `brew bundle dump --global` dumps it in `${HOME}/.Brewfile`. If you have to completely reinstall Homebrew (whatever reason) _and_ want to reinstate all previously installed packages:
```
# Dump your current Homebrew state
brew bundle dump --global

# Do Homebrew reinstall
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

# Restore your previous Homebrew state
brew bundle --global
```

**Reply by aviexk** [reply]
exactly, even if brew removes all the logs and caches, the executables still remain on my system. so in order to "re-index" them in brew, one has to install each one using brew again. 

further, brew doesn't even recognise the previous installed versions, like it would otherwise when trying to install something already installed. 
```❯ brew install zsh
Warning: zsh 5.8_1 is already installed and up-to-date.
To reinstall 5.8_1, run:
  brew reinstall zsh
```

**Reply by gromgit** [reply]
> exactly, even if brew removes all the logs and caches, the executables still remain on my system.

Using the official uninstaller _wipes everything Homebrew-related_. I know this to be true, because I just ran it:
```
# I installed `coreutils`, which contains at least the following...
$ which gtail
/usr/local/opt/coreutils/bin/gtail

$ ls -l /usr/local/opt/coreutils
lrwxr-xr-x 1 aho admin 23 Dec  2 12:08 /usr/local/opt/coreutils -> ../Cellar/coreutils/9.0/

$ ls -l /usr/local/Cellar
total 488464
-rw-r--r-- 1 aho admin         0 Dec 30  2020  .keepme
[...]
drwxr-xr-x 3 aho admin        96 Sep 26 23:21  coreutils/
[...]

# Now let's uninstall Homebrew...
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"
Warning: This script will remove:
/Users/aho/Library/Caches/Homebrew -> /Users/aho/Library/Caches/Homebrew
/Users/aho/Library/Logs/Homebrew/
/Volumes/DATA/Homebrew/Caches/
/usr/local/Caskroom/
/usr/local/Cellar/
/usr/local/bin/brew -> /usr/local/bin/brew
Are you sure you want to uninstall Homebrew? This will remove your installed packages! [y/N] y
==> Removing Homebrew installation...
Warning: Failed to delete /usr/local/Caskroom
rm: /usr/local/Caskroom: Permission denied
Warning: Failed to delete /usr/local/Cellar
rm: /usr/local/Cellar: Permission denied
==> Removing empty directories...
Password:
==> /usr/bin/sudo /usr/bin/find /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/opt /usr/local/sbin /usr/local/share /usr/local/var /usr/local/Caskroom /usr/local/Cellar /usr/local/Homebrew /usr/local/Frameworks -name .DS_Store -delete
==> /usr/bin/sudo /usr/bin/find /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/opt /usr/local/sbin /usr/local/share /usr/local/var /usr/local/Caskroom /usr/local/Cellar /usr/local/Homebrew /usr/local/Frameworks -depth -type d -empty -exec rmdir {} ;
==> Homebrew uninstalled!
The following possible Homebrew files were not deleted:
/usr/local/.com.apple.installer.keep
/usr/local/Frameworks/
/usr/local/Homebrew/
/usr/local/bin/
/usr/local/etc/
/usr/local/homebrew.rc
/usr/local/include/
/usr/local/lib/
/usr/local/opt/
/usr/local/sbin/
/usr/local/share/
/usr/local/var/
You may wish to remove them yourself.

# And see if coreutils still exists...
$ ls -l /usr/local/opt/coreutils
ls: /usr/local/opt/coreutils: No such file or directory

$ ls -l /usr/local/Cellar
ls: /usr/local/Cellar: No such file or directory

$ which gtail
# Nothing is found
```
The `Cellar` directory above is where Homebrew puts all the files from every `brew install`ed package.

A subsequent install of Homebrew creates a new _empty_ Cellar:
```
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Checking for `sudo` access (which may request your password)...
Password:
==> This script will install:
/usr/local/bin/brew
/usr/local/share/doc/homebrew
/usr/local/share/man/man1/brew.1
/usr/local/share/zsh/site-functions/_brew
/usr/local/etc/bash_completion.d/brew
/usr/local/Homebrew
==> The following existing directories will be made writable by user only:
/usr/local/share/zsh
/usr/local/share/zsh/site-functions
==> The following new directories will be created:
/usr/local/Cellar
/usr/local/Caskroom

Press RETURN to continue or any other key to abort:
==> /usr/bin/sudo /bin/chmod u+rwx /usr/local/share/zsh /usr/local/share/zsh/site-functions
==> /usr/bin/sudo /bin/chmod go-w /usr/local/share/zsh /usr/local/share/zsh/site-functions
==> /usr/bin/sudo /bin/mkdir -p /usr/local/Cellar /usr/local/Caskroom
==> /usr/bin/sudo /bin/chmod ug=rwx /usr/local/Cellar /usr/local/Caskroom
==> /usr/bin/sudo /usr/sbin/chown aho /usr/local/Cellar /usr/local/Caskroom
==> /usr/bin/sudo /usr/bin/chgrp admin /usr/local/Cellar /usr/local/Caskroom
==> /usr/bin/sudo /usr/sbin/chown -R aho:admin /usr/local/Homebrew
==> /usr/bin/sudo /bin/mkdir -p /Users/aho/Library/Caches/Homebrew
==> /usr/bin/sudo /bin/chmod g+rwx /Users/aho/Library/Caches/Homebrew
==> /usr/bin/sudo /usr/sbin/chown -R aho /Users/aho/Library/Caches/Homebrew
==> Downloading and installing Homebrew...
remote: Enumerating objects: 189, done.
remote: Counting objects: 100% (189/189), done.
remote: Compressing objects: 100% (115/115), done.
remote: Total 189 (delta 84), reused 157 (delta 67), pack-reused 0
Receiving objects: 100% (189/189), 100.40 KiB | 4.78 MiB/s, done.
Resolving deltas: 100% (84/84), completed with 13 local objects.
From https://github.com/Homebrew/brew
   c75aaed83..5096d6eab  master                                                   -> origin/master
 * [new branch]          dependabot/bundler/Library/Homebrew/rubocop-sorbet-0.6.3 -> origin/dependabot/bundler/Library/Homebrew/rubocop-sorbet-0.6.3
Checking out files: 100% (2720/2720), done.
HEAD is now at 5096d6eab Merge pull request #12515 from Homebrew/dependabot/bundler/Library/Homebrew/tapioca-0.5.6
Updated 2 taps (homebrew/core and homebrew/cask).
==> Installation successful!

==> Homebrew has enabled anonymous aggregate formulae and cask analytics.
Read the analytics documentation (and how to opt-out) here:
  https://docs.brew.sh/Analytics
No analytics data has been sent yet (nor will any be during this install run).

==> Homebrew is run entirely by unpaid volunteers. Please consider donating:
  https://github.com/Homebrew/brew#donations

==> Next steps:
- Run brew help to get started
- Further documentation:
    https://docs.brew.sh

$ ls -ld /usr/local/Cellar
drwxrwxr-x  2 aho  admin  64 Dec  5 18:21 /usr/local/Cellar

$ ls -l /usr/local/Cellar
# Nothing to see here
```
You mentioned in a separate thread that you're using an M1 Mac, so the official uninstaller will wipe everything Homebrew from whatever architecture-specific hierarchy you ran the uninstaller in (`/usr/local` for Rosetta installations, `/opt/homebrew` for native Apple Silicon). Therefore, one thing that could explain what you're seeing, if you're using the official Homebrew install/uninstall scripts, is that you initially installed Homebrew for _both_ architectures, then uninstalled one of them (which would wipe everything under that architecture) and continued using the other.

Or, as @SMillerDev mentioned, you installed and uninstalled Homebrew as one user, then reinstalled it as a different user.

Whatever the case, I've just demonstrated above that what you described simply cannot happen _as you described it_, so there are clearly some details that you left out. A normal uninstall, completed without errors, would wipe everything you previously installed via Homebrew, so there's nothing left to "reindex".

**Reply by aviexk** [reply]
okay maybe, I haven't been clear about the issue. Apologies for that. I'm going to give it another shot below, @SMillerDev maybe this also helps our thread. 

More specifically, I'm referring to the macOS native applications(Casks). (The `discord.app` executables created in the `/Applications` directory on macOS upon calling `brew install discord`.)

This is something more related to `brew cask` maybe? idk

So to reiterate, 
1. install brew itself (using official installer)
2. `brew install discord`(or any Cask)
3. `brew list` returns discord
4. uninstall brew (for whatever reason) (using official uninstaller)
5. reinstall brew (for whatever reason) (using official instalelr)
6. brew list returns nothing

**but the executable is still there in `/Applications' and the app runs.**

1) either uninstalling brew should also presumably remove the files it installs from '/Applications' directory 
2) and since it doesnt, there is no way to tell brew to install those casks again(this is what I meant by the "reindexing" earlier)

**Reply by SMillerDev** [reply]
Casks are special in that they aren't brew packages. They are packages from other developers that brew just moves or runs the installer off. 

If you remove all brew data it will remove information about what installs were ran for casks. The only way to regain that information is to run the installs again.

**Comment by SMillerDev** [comment]
Do you have an M1 processor and is your old install based on Rosetta?

**Reply by aviexk** [reply]
yes running on apple silicon, using brew for packages of both architectures (arm64 and X86)

**Reply by SMillerDev** [reply]
Well, those are separate installs.  So you'd have to run the installer with and without Rosetta and call brew with and without Rosetta to manage the packages.

**Reply by aviexk** [reply]
yes, I'm aware of the separation of the two architecture based brews running on my systems. my issue wasn't with the separate installs, and nothing specific to m1 chips. 

My issue still remains, after brew is uninstalled I still have the executables on my system, and those are unrecognised by brew once I reinstall it, so essentially i have to find out everything that was installed and attempt to re-install the package with brew again one-by-one to manage them under brew again.

is there anyway around this?

**Reply by SMillerDev** [reply]
Since brew uses the files on disk to know what was installed there is only one possible cause of this. Brew can't find the files, meaning they are not accessible to the user brew runs as or brew is looking in the wrong directory (most likely caused by a Rosetta/native difference)

**Reply by aviexk** [reply]
https://github.com/Homebrew/discussions/discussions/2580#discussioncomment-1752949
